File: C++/0912-sort-an-array.cpp
class Solution {
private:
    void swap(int& a, int& b){
        int tmp = a;
        a = b;
        b = tmp;
    }
	void merge(vector<int>& nums, int l, int m, int r){
        vector<int> tmp(r - l + 1);
        int i = l, j = m + 1, k = 0;
        while(i <= m && j <= r){
            if(nums[i] <= nums[j])
                tmp[k++] = nums[i++]; 
            else
                tmp[k++] = nums[j++];
        }
        while(i <= m) tmp[k++] = nums[i++];
        while(j <= r) tmp[k++] = nums[j++]; 
        for(i = 0; i < k; i++)
            nums[l + i] = tmp[i];
    }
	
    void merge_sort(vector<int>& nums, int l, int r){
        if(l >= r) 
            return;
        int m = l + (r - l) / 2;
        merge_sort(nums, l, m);
        merge_sort(nums, m + 1, r);
        merge(nums, l, m, r);
    }
public:
    vector<int> sortArray(vector<int>& nums) {
        merge_sort(nums, 0, nums.size() - 1);
        return nums;
    }
};

File: Python/0912-sort-an-array.py
class Solution:
    def partition(self, nums, left, right):
        tmp = nums[right]
        i = left
        for j in range(left, right):
            if nums[j] < tmp:
                nums[i], nums[j] = nums[j], nums[i]
                i = i +1
        nums[i], nums[right] = nums[right], nums[i]
        return i
    def quicksort(self, nums, left, right):
        if left > right:
            return
        i = self.partition(nums, left, right)
        self.quicksort(nums, left, i -1)
        self.quicksort(nums, i+1, right)
    def sortArray(self, nums: List[int]) -> List[int]:
        if len(nums) <= 1:
            return nums
        self.quicksort(nums, 0, len(nums) - 1)
        return nums
